### PR TITLE
List

### PR DESCRIPTION
**Pull request by mettin**

### CODE REVIEW

File: index.jsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import List from './src/components/List';
+
+import cx from 'classnames';
+
+function ListContainer(props) {
+	return (
```
Comment (mettin) [inline index.jsx]: Using tabs? damn

File: src/components/List.jsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import cx from 'classnames';
+import _ from 'lodash';
```
Comment (mettin) [inline src/components/List.jsx]: Heb je de hele lodash lib nodig? kun je niet een custom buildje maken met alleen de 2-3 dingen die je gebruikt?

File: src/components/List.jsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+	getDefaultProps() {
+		return {
+			element: 'ul'
+		};
+	},
+
+	renderChildren() {
+		const items = this.props.items;
+		let listItem = this.props.children;
```
Comment (mettin) [inline src/components/List.jsx]: const?

File: src/components/List.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+		return items.map(function generateItem(item) {
```
Comment (mettin) [inline src/components/List.jsx]: move up generateItem
```js
return items.map(this.generateItem);
```

File: src/components/List.jsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+			if (! listItem) {
+				listItem = <li className="list-group-item">{item}</li>;
+			}
+
+			return React.Children.map(listItem, function(child) {
+				const props = Object.assign({}, child.props);
+
+				if (typeof props.children === 'string') {
+					return ('' + props.children).replace(/(\$item)(\..*)*/, (match, p1, p2, offset, origin) => {
```
Comment (mettin) [inline src/components/List.jsx]: Whoa, can't read. move regex to util

File: src/components/List.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		const props = Object.assign({
+			className: cx('list-group')
+		}, this.props);
+		const element = props.element;
+
+		delete props.items;
+		delete props.element;
+
+		return React.createElement(element, props, this.renderChildren(props.items));
+			// <element {...props}>
```
Comment (mettin) [inline src/components/List.jsx]: remove comments

File: test/components/Icon_spec.jsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+			this.renderer.render(icon);
+			this.DOM = this.renderer.getRenderOutput();
+		});
+
+		it('is a normal FontAwesome icon by default', function() {
+			const icon = this.DOM;
+
+			expect(icon.type).to.equal('i');
+			expect(icon.props.className.indexOf('fa') > -1).to.be.true;
```
Comment (mettin) [inline test/components/Icon_spec.jsx]: i think you can do
```js
expect(icon.props.classList.contains('fa')).to.be.true;
```

File: test/components/Icon_spec.jsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		this.renderer = TestUtils.createRenderer();
+	});
+
+	describe('rendering', function() {
+
+		beforeEach(function() {
+			const icon = <Icon icon="plus" title="Add more" />;
+
+			this.renderer.render(icon);
+			this.DOM = this.renderer.getRenderOutput();
```
Comment (mettin) [inline test/components/Icon_spec.jsx]: this.DOM implies the entire document to me. change to this.iconElement

File: test/components/Icon_spec.jsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+			expect(icon.props.className.indexOf('fa-plus') > -1).to.be.true;
+			expect(icon.props.className.indexOf('fa-fw') === -1).to.be.true;
+		});
+
+		it('is decorative', function() {
+			expect(this.DOM.props['aria-hidden']).to.be.true;
+		});
+
+		it('accepts additional attributes', function() {
+			const icon = this.DOM;
```
Comment (mettin) [inline test/components/Icon_spec.jsx]: why create single use const?

File: test/components/List_spec.jsx [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import {expect} from 'chai';
+import TestUtils from 'react-addons-test-utils';
+import List from '../../src/components/List';
+
+describe('the List component', function() {
```
Comment (mettin) [inline test/components/List_spec.jsx]: Add test for changing element tag

File: src/components/List.jsx (hunk 2 above)
Comment (remypar5) [inline src/components/List.jsx]: mettin: 

> "Do you really need the full lodash lib? Wouldn't it be better to create a filtered custom build?"

I guess that's a good idea. Will do that in one of the next iterations

File: dist/index.html [hunk 11]
```diff
@@ -3,6 +3,8 @@
 <head>
 	<title>JACl test page</title>
 	<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
+	<!-- Latest compiled and minified CSS -->
+	<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
```
Comment (mettin) [inline dist/index.html]: include in project, loading from cdn might fail.

File: index.jsx [hunk 12]
```diff
@@ -1,18 +1,24 @@
...
-		<List items={props.items}>
-			<div className="list-group-item">item: $item.text.text</div>
-		</List>
-	);
-};
+const classes = ['warning', 'info', 'success', 'default'];
+
+const ListContainer = React.createClass({
+	generateItem(item, key, ctx) {
+		const className = cx('list-group-item', 'list-group-item-' + classes[key % classes.length]);
```
Comment (mettin) [inline index.jsx]: isnt it `list--group-item__${modifier}`
`block--element__modifier`

File: index.jsx [hunk 13]
```diff
@@ -1,18 +1,24 @@
...
+		return <div key={key} className={className}>item: {item}</div>;
```
Comment (mettin) [inline index.jsx]: *personal preference*
when there are more then 1 attr, newline per attr

File: src/components/List.jsx [hunk 14]
```diff
@@ -10,56 +10,26 @@ export default React.createClass({
...
-			// 	})}
-			// </element>
-		// );
-	}
+	generateItem(item, key, ctx) {
+		return <li key={key} className="list-group-item">{item}</li>;
+	},
+
+	renderChildren() {
+		const generator = this.props.generateItem ? this.props.generateItem : this.generateItem;
```
Comment (mettin) [inline src/components/List.jsx]: also check if `typeof this.props.generateItem === 'function'`

File: src/components/List.jsx [hunk 15]
```diff
@@ -10,56 +10,26 @@ export default React.createClass({
...
+		return _.map(this.props.items, generator, this);
```
Comment (mettin) [inline src/components/List.jsx]: why don't you use native js .map?

File: test/components/List_spec.jsx [hunk 16]
```diff
@@ -3,26 +3,53 @@ import {expect} from 'chai';
 import TestUtils from 'react-addons-test-utils';
 import List from '../../src/components/List';
 
-describe('the List component', function() {
+describe('<List />', function() {
```
Comment (mettin) [inline test/components/List_spec.jsx]: convenience tip: 
`describe('test/components/List_spec.jsx', function() {`
so when there's an error you can see the file where it's coming from

File: test/components/List_spec.jsx [hunk 17]
```diff
@@ -3,26 +3,53 @@ import {expect} from 'chai';
...
+		it('renders a <li> element per item', function() {
+			const list = renderComponent(
+				<List items={['item1', 'item2']} />
+			);
+
+			const listItems = TestUtils.scryRenderedDOMComponentsWithTag(list, 'li');
+
+			expect(listItems.length).to.equal(2);
+			expect(listItems[0].tagName).to.equal('LI');
+			expect(listItems[0].className.indexOf('list-group-item') >= -1).to.be.true;
```
Comment (mettin) [inline test/components/List_spec.jsx]: http://chaijs.com/api/bdd/#method_include
```js
expect(listItems[0].className).to.contain('list-group-item');`
```

File: src/components/List.jsx (hunk 14 above)
Comment (remypar5) [inline src/components/List.jsx]: I'll leave that to `this.propTypes`

File: test/components/List_spec.jsx (hunk 10 above)
Comment (remypar5) [inline test/components/List_spec.jsx]: And also for passing in a hashmap (object) as this.props.items

**Pull request merged**